src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -10505,6 +10577,13 @@ private TypeWithAnnotations GetDeclaredParameterResult(ParameterSymbol parameter
             var resultType = ApplyUnconditionalAnnotations(indexer.TypeWithAnnotations.ToTypeWithState(), GetRValueAnnotations(indexer));
             SetResult(node, resultType, indexer.TypeWithAnnotations);
             SetUpdatedSymbol(node, node.Indexer, indexer);
+
+            if (node.Type.IsDynamic() && !node.Indexer.Type.IsDynamic())
+            {
+                Debug.Assert(!node.Indexer.ReturnsByRef);
+                ForceDynamicResultType(node, node.Indexer.Type);
+            }
```

Comment:
Consider doing `ForceDynamicResultType` OR `SetResult`, but not both. Applies here and elsewhere in nullable walker.
```
            if (node.Type.IsDynamic() && !node.Indexer.Type.IsDynamic())
            {
                Debug.Assert(!node.Indexer.ReturnsByRef);
                ForceDynamicResultType(node, node.Indexer.Type);
            }
            else 
            {
                SetResult(node, resultType, indexer.TypeWithAnnotations);
            }
```

Then `ForceDynamicResultType` can be renamed to `SetDynamicResult`.